### ISSUE TITLE
podman top: implement ps args

### ISSUE THREAD

**Issue by edsantiago** [projectatomic/libpod#1006]
`podman top` ignores ps OPTIONS:
```go
libpod/container_top.go
       // [...] We believe that
       // writing a go library which parses procFS in a ps-compatible way may
       // be more beneficial in the long run.  Until then, args will be
       // ignored.
```
This is a simple reminder so the issue is not forgotten.

(And ITM someone might want to upodate the man page, especially if the fix will take a long time)

**Comment by vrothberg** [comment]
I am planning to write a golang-lib to parse procFS, so we don't depend on `ps (1)` anymore. Other projects like Moby but also runc would benefit from it as well.

**Comment by mheon** [comment]
Agree with a manpage update being a good idea. Probably should have asked for that when I requested that we disable that arg. Oops.

**Comment by cgwalters** [comment]
This is only tangentially related but...the way pids are enumerated/managed looks racy, part of it is the cgroup API's fault.  Basically due to possible pid reuse, what we read from `/sys/fs/cgroup/pids/.../tasks` may be out of date by the time we go to invoke `ps` or read from `/proc`.  The way polkit handles this is to use the pair (pid, start-time) as a unique process identifier.  See [the code](https://cgit.freedesktop.org/polkit/tree/src/polkit/polkitunixprocess.c?id=373705b35e7f6c7dc83de5e0a3ce11ecd15d0409#n56).

I think to make this saner a kernel patch to add `/sys/fs/cgroup/pids/.../tasks-and-starttime` or so would help.  Or perhaps in the meantime, go back and verify the pid is still in the cgroup after gathering the data from `/proc`.

(Though I didn't actually verify this race exists, just was reading the code)

**Comment by vrothberg** [comment]
> I am planning to write a golang-lib to parse procFS, so we don't depend on ps (1) anymore. Other projects like Moby but also runc would benefit from it as well.

I started working on this yesterday

**Comment by rhatdan** [comment]
Lets get this library under github.com/containers/psgo and then we can work on it there.

**Comment by vrothberg** [comment]
> Library is now in containers/psgo, @vrothberg Could you work on this issue?

@rhatdan, with pleasure. Let's figure out some remaining issues of containers/psgo first and then I can add that to libpod.

**Comment by vrothberg** [comment]
This issue has been fixed with commit ba1871dac033783ab0329c9b3c9113a34a90992f.
